### PR TITLE
reworked image loading/overwriting

### PR DESCRIPTION
**Pull request by phisch**
## Description
Since we had problems overwriting favicons in themes, i took a closer look at the existing code for that functionality. Turned out it was clustered with several edge cases and included non functional legacy workarounds. I took some time to re-work that piece of code to harden the implementation of that feature and make it more readable in general.

Themes are now able to overwrite the following images:
 - images from apps, even from apps with special locations
 - images from core "apps" like core or settings

I threw out all the svg checks, since they did not make any sense. Overwriting and falling back from any image extension to png files still works, but the requested file will be treated preferred.

## Related Issue
https://github.com/owncloud/enterprise/issues/1832

## Motivation and Context
It solves mainly the error where some files could not be overwritten. (like favicons)

## How Has This Been Tested?
Has been tested by hand with example and enterprise themes.
Image loading will be tested in a reasonable extend. It's really hard to provide solid test cases for all the possibilities this provides. This has not been tested before either, i suspect that was because of the same reasons.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by phisch** [comment]
Can be reviewed already. Like said in the original comment, some tests will be added, but most of the scenarios are really hard to test.

**Comment by PVince81** [comment]
It might be possible to test this with integration tests and an example theme that overrides various images. The integration test would need to request the overridden images through HTTP and see if they exist (no 404).

@phisch are we losing anything else apart from non-functional legacy fallbacks ? It's quite a big chunk of code that got deleted.

**Comment by phisch** [comment]
@PVince81 except for the weird svg behavior it should behave the same (just more stable). The png fallback is now a general thing instead of just in some cases, and overwriting files should be straight forward. I had legacy themes in mind while developing that. Also i tested all themes i could find and i did not see anything wrong and all requests to images where successful.

Your idea with the integration tests is good, i will have a look at them.

File: lib/private/URLGenerator.php [hunk 1]
```diff
@@ -60,6 +62,7 @@ public function __construct(IConfig $config,
 		$this->config = $config;
 		$this->cacheFactory = $cacheFactory;
 		$this->router = $router;
+		$this->theme = \OC_Util::getTheme();
```
Comment (DeepDiver1975) [inline lib/private/URLGenerator.php]: inject ?

**Comment by PVince81** [comment]
Fine then :+1:

**Pull request merged**